src/abi/solana_accounts.rs

```diff
@@ -0,0 +1,414 @@
+                Sign::Plus,
+                &"SysvarRent111111111111111111111111111111111"
+                    .from_base58()
+                    .unwrap(),
+            ),
+            SolanaAccount::Rent,
+        ),
+        (
+            BigInt::from_bytes_be(
+                Sign::Plus,
+                &"TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
+                    .from_base58()
+                    .unwrap(),
+            ),
+            SolanaAccount::TokenProgramId,
+        ),
+        (
+            BigInt::from_bytes_be(
+                Sign::Plus,
+                &"SysvarC1ock11111111111111111111111111111111"
+                    .from_base58()
+                    .unwrap(),
+            ),
+            SolanaAccount::ClockAccount,
+        ),
+    ])
+});
+
+impl SolanaAccount {
+    /// Retrieve a name from an account, according to Anchor's constant accounts map
+    /// https://github.com/coral-xyz/anchor/blob/06c42327d4241e5f79c35bc5588ec0a6ad2fedeb/ts/packages/anchor/src/program/accounts-resolver.ts#L54-L60
+    pub(super) fn name(&self) -> String {
+        match self {
+            SolanaAccount::AssociatedProgramId => "associatedTokenProgram".to_string(),
+            SolanaAccount::Rent => "rent".to_string(),
+            SolanaAccount::SystemAccount => "systemProgram".to_string(),
+            SolanaAccount::TokenProgramId => "tokenProgram".to_string(),
+            SolanaAccount::ClockAccount => "clock".to_string(),
+            SolanaAccount::InstructionAccount => "SysvarInstruction".to_string(),
+        }
+    }
+
+    fn from_number(num: &BigInt) -> Option<SolanaAccount> {
+        AVAILABLE_ACCOUNTS.get(num).cloned()
+    }
+}
+
+/// Struct to save the recursion data when traversing all the CFG instructions
+struct RecurseData<'a> {
+    /// Here we collect all accounts. It is a map between the cfg number of a functions and the
+    /// accounts it requires
+    accounts: HashMap<usize, IndexSet<SolanaAccount>>,
+    /// next_queue saves the set of functions we must check in the next iteration
+    next_queue: IndexSet<usize>,
+    /// The number of the function we are currently traversing
+    cfg_func_no: usize,
+    /// The quantity of accounts we have added the the hashmap 'accounts'
+    accounts_added: usize,
+    ns: &'a Namespace,
+}
+
+impl RecurseData<'_> {
+    /// Add an account to the hashmap
+    fn add_account(&mut self, account: SolanaAccount) {
+        let inserted = if let Some(set) = self.accounts.get_mut(&self.cfg_func_no) {
+            set.insert(account)
+        } else {
+            self.accounts
+                .insert(self.cfg_func_no, IndexSet::from([account]));
+            true
+        };
+        if inserted {
+            self.accounts_added += 1;
+        }
+    }
+}
+
+/// Collect the accounts this contract needs
+pub(super) fn collect_accounts_from_contract(
+    contract_no: usize,
+    ns: &Namespace,
+) -> HashMap<usize, IndexSet<SolanaAccount>> {
```

Comment:
Also, `Instr::ExternalCall` does not tell me which contract I am calling.

Reply:
Another thing to add that I've just remembered and I have discussed this with the Anchor folks. When doing an external call, the developer must write in Solidity the AccountMeta specifying the accounts for the transaction. This should not be done in Typescript (this is what the Anchor folks told me).

Reply:
> Another thing to add that I've just remembered and I have discussed this with the Anchor folks. When doing an external call, the developer must write in Solidity the AccountMeta specifying the accounts for the transaction. This should not be done in Typescript (this is what the Anchor folks told me).

I am not following. 

You don't do external calls in typescript. You build a transaction in typescript. external calls/CPI is purely an on-chain thing.

Reply:
> You are right. I can do the best effort to collect the accounts for other contracts. I have not implemented this because we can call another contract through external call and don't have its functions definitions. This is the case when we parse the IDL from either an Anchor contract or another Solang Solidity contract deployed on chain.

Ideally if we call another function through parsed idl, then the accounts will be populated too.